src/modules/claims/claims.service.ts

```diff
@@ -261,28 +262,61 @@ export class ClaimsService {
     return this._cacheClient.issueClaim(this._signerService.did, message);
   }
 
+  /**
+   *
+   * @param token optional token containing claimType, version and subject
+   * @returns claim params obtained from token
+   */
+  private getClaimTypeFromToken(token?: string) {
+    if (token) {
+      const { claimData, sub } = this._didRegistry.jwt.decode(token) as {
+        claimData: { claimType: string; claimTypeVersion: string };
+        sub: string;
+      };
+      return { ...claimData, subject: sub };
+    }
+    return {};
+  }
+
   /**
    * @description Registers issued onchain claim with Claim manager
    *
-   * @param claimId - id of signed onchain claim
+   * @param claim - id of signed onchain claim.
+   * @param token - @deprecated use subject, claimType, claimTypeVersion instead. Token should get removed and subject, claimType, claimTypeVersion should be required.
    */
-  async registerOnchain(
-    claim: Pick<Claim, 'token' | 'subjectAgreement' | 'onChainProof' | 'acceptedBy'>
-  ) {
-    if (!readyToBeRegisteredOnchain(claim)) {
-      throw new Error(ERROR_MESSAGES.CLAIM_WAS_NOT_ISSUED);
+  async registerOnchain(claim: {
+    claimType?: string;
+    claimTypeVersion?: string;
+    token?: string;
+    subjectAgreement?: string;
+    onChainProof?: string;
+    acceptedBy?: string;
+    subject?: string;
+  }) {
+    // backward compatibility with token
+    claim = { ...claim, ...this.getClaimTypeFromToken(claim.token) };
+
+    if (
+      !claim.subjectAgreement &&
+      claim.subject === this._signerService.did &&
+      claim.claimType &&
```

Comment:
We can ensure that `claimType`, `claimTypeVersion` and `subject` are set at this point like this
```
if (!claimType || !...) {
   ({claimType, ...} = this.getClaimTypeFromToken(claim.token));
}
```

Reply:
it won't be sure since claim.token is optional. That's why I use token if it exists and then I validate against all required params existance.

Reply:
if (!claimType || !claimTypeVersion || !subject) and !token then error should be thrown

Reply:
Existance of 'claimType', 'claimTypeVersion', 'subject', 'onChainProof', 'acceptedBy', 'subjectAgreement' is validated later on. Before then we are trying to fill the gaps : 'claimType', 'claimTypeVersion', 'subject' from token if possible and 'subjectAgreement' if not provided. 
Then we validate whole request.